qiskit/circuit/qpy_serialization.py

```diff
@@ -936,8 +1024,96 @@ def _write_instruction(file_obj, instruction_tuple, custom_instructions, index_m
         container.close()
 
 
+def _write_pauli_evolution_gate(file_obj, evolution_gate):
+    operator_list = evolution_gate.operator
+    standalone = False
+    if not isinstance(operator_list, list):
+        operator_list = [operator_list]
+        standalone = True
+    num_operators = len(operator_list)
+    pauli_data_buf = io.BytesIO()
+    for operator in operator_list:
+        element_buf = io.BytesIO()
+        buf = io.BytesIO()
+        pauli_list = operator.to_list(array=True)
+        np.save(buf, pauli_list)
+        data = buf.getvalue()
+        element_metadata = struct.pack(SPARSE_PAULI_OP_LIST_ELEM_PACK, len(data))
+        element_buf.write(element_metadata)
+        element_buf.write(data)
+        pauli_data_buf.write(element_buf.getvalue())
```

Comment:
Should we either be re-using our buffers or closing them at the end of each loop iteration?  I have a suspicion I've seen Python complaining about garbage collecting these objects without them having been closed, but I couldn't spot it in the docs immediately.

If you don't know what I'm talking about, we can probably just ignore this as a false memory of mine.

Reply:
I'm not sure about reusing buffers I get worried about leaking data between them causing weird bugs. I have a vague memory about something like that too but I don't have any specifics either. I've always found the docs about the `io` module a bit lacking. Like I think we can use a context manager here to ensure we close it but I can't find that in the docs. Anyway I'll update this to ensure we close things to just make sure this isn't a memory leak

Reply:
Oh yeah, good call about buffer overrun bugs.  Let's not re-use them - that was a bad suggestion.  Close and re-open seems safe and should help out the gc at the very least.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7374/commits/82d3576f13866eb006314ce338da7ef54b0053d6